File: 1DOMACI/4th_problem_Opencv/4th_problem_OpenCV.cpp

/*
 * 4. Napisati funkciju koja učitava RGB sliku po ličnom izboru i prikazuje informacije o
 * slici. Nakon toga razdvojiti R, G i B komponente slike i prikazati sve slike istovremeno
 * u jednom prozoru. Komponentu po izboru sačuvati kao intenzitetnu sliku. Početnu RGB sliku
 * konvertovati u indeksnu, prikazati je i sačuvati kao novu sliku. Prikazati polaznu sliku
 * kao sliku u ogledalu originalne slike po proizvoljno izabranoj osi i naznačiti koja je
 * osa izabrana.
*/

#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>
// #include <libexif/exif-data.h>  // <- TODO: see how to set up cmake
// #include <libexif/exif-loader.h>

#define NUMCOLORS 12

// Namespace nullifies the use of cv::function(); 
using namespace std;
using namespace cv;

int main(){ 
    
    // Load image
    // ----------------------------
    Mat img = imread("../MrBean.jpg", IMREAD_UNCHANGED);


    // Show image
    // ----------------------------
    imshow("Original image", img);  
    waitKey(0);                       
    destroyAllWindows();
    

    // Metadata
    // ----------------------------
    // TODO: Set up for use with exif.
    std::cout << "IMAGE INFORMATION" << std::endl
              << "Dimensions: (" << img.cols << " x " << img.rows << ")" << std::endl
              << "Metadata using OpenCV here is limited. libexif can be used, but "
              << "I need to figure out how to set up cmake for it." << std::endl;


    // Color channels
    // ----------------------------
    // Showing all images in the same window is a tedious task and I shall not do it here.
    string colors[3] = {"Blue", "Green", "Red"};
    Mat bgr[3];
    split(img, bgr);
    for (int color = 0; color < 3; color++) {
        imshow(colors[color], bgr[color]);
        waitKey(0);                       
        destroyAllWindows();
        imwrite(colors[color] + ".jpg", bgr[color]);
    }


    // Indexed image
    // ----------------------------
    /*
     * TODO: Write index conversion.
    */

    return 0;
}

File: 1DOMACI/4th_problem_Opencv/4th_problem_OpenCV.py
'''
 * 4. Napisati funkciju koja učitava RGB sliku po ličnom izboru i prikazuje informacije o
 * slici. Nakon toga razdvojiti R, G i B komponente slike i prikazati sve slike istovremeno
 * u jednom prozoru. Komponentu po izboru sačuvati kao intenzitetnu sliku. Početnu RGB sliku
 * konvertovati u indeksnu, prikazati je i sačuvati kao novu sliku. Prikazati polaznu sliku
 * kao sliku u ogledalu originalne slike po proizvoljno izabranoj osi i naznačiti koja je
 * osa izabrana.
'''


# Using OpenCV

import cv2 as cv
from matplotlib import pyplot as plt
import numpy as np

def close_event():
    plt.close()
    print('CLOSED')

if __name__ == '__main__':

    # Load image
    img = cv.imread('IMGS/MrBean.jpg', cv.IMREAD_COLOR)

    # Show original image
    plt.figure(0)
    plt.axis('off')
    plt.imshow(cv.cvtColor(img, cv.COLOR_BGR2RGB))
    plt.title('ORIGINAL')
    plt.draw()
    #plt.show()
    # NOTE: For some reason, cv.imshow freezes my console.

    # Split channels
    B, G, R = cv.split(img)
    channels = {1:R, 2:G, 3:B}
    names = {1:'RED', 2:'GREEN', 3:'BLUE'}
    plt.figure(1)
    for key in channels.keys():
        plt.subplot(1,3,key)
        plt.axis('off')
        plt.imshow(channels[key], cmap='gray')
        plt.title(names[key])
    #plt.show()

    # Save channel as intensity
    plt.figure(2)
    plt.imshow(channels[1], cmap='gray')
    plt.axis('off')
    plt.savefig('IMGS/red_chl_py.jpg')

    # Image quantization
    # Code taken from https://stackoverflow.com/questions/11064454/adobe-photoshop-style-posterization-and-opencv
    # This is great python code...
    n = 4
    indices = np.arange(0,256)
    divider = np.linspace(0,255,n+1)[1]
    quantiz = np.int0(np.linspace(0,255,n))
    color_levels = np.clip(np.int0(indices/divider),0,n-1)
    palette = quantiz[color_levels]             # B E A U T I F U L !!!
    img_quantized = palette[img]                # Palette method is faster and shorter!!!
    img_quantized = cv.convertScaleAbs(img_quantized)

    plt.figure(3)
    plt.axis('off')
    plt.imshow(cv.cvtColor(img_quantized, cv.COLOR_BGR2RGB))
    plt.savefig('IMGS/img_quantized_py.jpg')
    plt.title('QUANTIZED')

    # Mirroring
    img_flipped_x = cv.flip(img, 0) # 0 -> x, >0 -> y, <0 -> xy

    plt.figure(4)
    plt.axis('off')
    plt.imshow(cv.cvtColor(img_flipped_x, cv.COLOR_BGR2RGB))
    plt.title('FLIPPED AROUND X')
    plt.show()
